### PR TITLE
Creates test scenarios and test zones for test cases Delegation01, -02 and -03

### PR DESCRIPTION
**Pull request by matsduf**
## Purpose

This PR defines test scenarios and create test zones for test cases
* Delegation01
* Delegation02
* Delegation03

New unit tests bases in this work have been created in https://github.com/zonemaster/zonemaster-engine/pull/1395.

The implementation of Delegation02 also outputs tags SAME_IP_ADDRESS or DISTINCT_IP_ADDRESS. Those are however not defined in the test case specification, and they are disregarded for the scenarios. They do not seem to add any information to the result.

The unit tests created from these scenarios will replace legacy unit tests.

## Context

https://github.com/zonemaster/zonemaster-engine/pull/1395

## How to test this PR

Review.

### CODE REVIEW

File: docs/public/specifications/test-zones/Delegation-TP/delegation01.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+NO-IPV4-1                     | ENOUGH_IPV6_NS_CHILD, ENOUGH_IPV6_NS_DEL, ENOUGH_NS_CHILD, ENOUGH_NS_DEL, NO_IPV4_NS_CHILD, NO_IPV4_NS_DEL | 2)
+NO-IPV4-2                     | ENOUGH_IPV6_NS_CHILD, ENOUGH_IPV6_NS_DEL, ENOUGH_NS_CHILD, ENOUGH_NS_DEL, NO_IPV4_NS_CHILD, NO_IPV4_NS_DEL | 2)
+NO-IPV4-3                     | ENOUGH_IPV6_NS_CHILD, ENOUGH_IPV6_NS_DEL, ENOUGH_NS_CHILD, ENOUGH_NS_DEL, NO_IPV4_NS_CHILD, NO_IPV4_NS_DEL | 2)
+NO-IPV6-1                     | ENOUGH_IPV4_NS_CHILD, ENOUGH_IPV4_NS_DEL, ENOUGH_NS_CHILD, ENOUGH_NS_DEL, NO_IPV6_NS_CHILD, NO_IPV6_NS_DEL | 2)
+NO-IPV6-2                     | ENOUGH_IPV4_NS_CHILD, ENOUGH_IPV4_NS_DEL, ENOUGH_NS_CHILD, ENOUGH_NS_DEL, NO_IPV6_NS_CHILD, NO_IPV6_NS_DEL | 2)
+NO-IPV6-3                     | ENOUGH_IPV4_NS_CHILD, ENOUGH_IPV4_NS_DEL, ENOUGH_NS_CHILD, ENOUGH_NS_DEL, NO_IPV6_NS_CHILD, NO_IPV6_NS_DEL | 2)
+MISMATCH-DELEGATION-CHILD-1   | ENOUGH_IPV4_NS_CHILD, NOT_ENOUGH_IPV4_NS_DEL, ENOUGH_IPV6_NS_CHILD, NOT_ENOUGH_IPV6_NS_DEL, ENOUGH_NS_CHILD, ENOUGH_NS_DEL | 2)
+MISMATCH-DELEGATION-CHILD-2   | NOT_ENOUGH_IPV4_NS_CHILD, ENOUGH_IPV4_NS_DEL, NOT_ENOUGH_IPV6_NS_CHILD, ENOUGH_IPV6_NS_DEL, ENOUGH_NS_CHILD, ENOUGH_NS_DEL | 2)
+
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/Delegation-TP/delegation01.md:86]: Fixed.

File: docs/public/specifications/test-zones/Delegation-TP/delegation02.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+ALL-DISTINCT-2                | DEL_DISTINCT_NS_IP, CHILD_DISTINCT_NS_IP | 2)
+ALL-DISTINCT-3                | DEL_DISTINCT_NS_IP, CHILD_DISTINCT_NS_IP | 2)
+DEL-NON-DISTINCT              | DEL_NS_SAME_IP, CHILD_DISTINCT_NS_IP     | 2)
+DEL-NON-DISTINCT-UND          | DEL_NS_SAME_IP, CHILD_DISTINCT_NS_IP     | 2)
+CHILD-NON-DISTINCT            | DEL_DISTINCT_NS_IP, CHILD_NS_SAME_IP     | 2)
+CHILD-NON-DISTINCT-UND        | DEL_DISTINCT_NS_IP, CHILD_NS_SAME_IP     | 2)
+NON-DISTINCT-1                | DEL_NS_SAME_IP, CHILD_NS_SAME_IP         | 2)
+NON-DISTINCT-2                | DEL_NS_SAME_IP, CHILD_NS_SAME_IP         | 2)
+NON-DISTINCT-3                | DEL_NS_SAME_IP, CHILD_NS_SAME_IP         | 2)
+
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/Delegation-TP/delegation02.md:68]: Fixed.

**Review by mattias-p — COMMENTED** [review]
I've only read through the markdown files and noted things I saw as I passed by. It's a big PR, but if you want I can go through it again more thoroughly.

File: docs/public/specifications/test-zones/Delegation-TP/delegation01.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+* (1) All tags except for those specified as "Forbidden message tags" (no
+  instances for these test scenarios)
+* (2) All tags except for those specified as "Mandatory message tags"
+
+## Zone setup for test scenarios
+
+Assumptions for the scenario specifications unless otherwise specified for the
+specific scenario:
+* For each scenario zone there are two name server configured.
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/Delegation-TP/delegation01.md]: Updated NS --> name server/name servers where applicable.

**Review by mattias-p — COMMENTED** [review]
I looked through the documentation again. I found a bunch of details to comment on, but overall it looks really good.

File: docs/public/specifications/test-zones/Delegation-TP/delegation02.md [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+* Zone: all-distinct-2.delegation02.xa
+  * Both ns1 and ns2 are out-of-bailiwick under the xb tree.
+    * ns1 is "ns1.all-distinct-2.delegation02.xb"
+    * ns2 is "ns2.all-distinct-2.delegation02.xb"
+  * Delegation is without glue.
+  * The test zone has no address records for the name server names.
+  * The "delegation02.xb" zone has full set of address records for this scenario.
+
+### ALL-DISTINCT-3
+This is also a happy path. Also out-of-bailiwick, but with sibbling glue.
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/Delegation-TP/delegation02.md]: Fixed all places.

File: docs/public/specifications/test-zones/Delegation-TP/delegation03.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+### REFERRAL-SIZE-TOO-LARGE-2
+Referal is too large and name servers are out-of-bailiwick with no glue.
+
+* Zone: referral-size-too-large-2.delegation03.xa
+  * The zone is delegated to ns1, ns2, ns3 and ns4.
+    * ns1 is "ns1.abcdefghijklmnopqrstuvwxyz.abcdefghijklmnopqrstuvwxyz.abcdefghijklmnopqrstuvwxy.referral-size-too-large-2.delegation03.xb"
+    * ns2 is "ns2.abcdefghijklmnopqrstuvwxyz.abcdefghijklmnopqrstuvwxyz.abcdefghijklmnopqrstuvwxy.referral-size-too-large-2.delegation03.xb"
+    * ns3 is "ns3.abcdefghijklmnopqrstuvwxyz.abcdefghijklmnopqrstuvwxyz.abcdefghijklmnopqrstuvwxy.referral-size-too-large-2.delegation03.xb"
+    * ns4 is "ns4.abcdefghijklmnopqrstuvwxyz.abcdefghijklmnopqrstuvwxyz.abcdefghijklmnopqrstuvwxy.referral-size-too-large-2.delegation03.xb"
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/Delegation-TP/delegation03.md:113]: Authoritative name servers might be able to use name compression for the second, third and fourth name server name. It’s probably better to follow a similar naming scheme as for the name server names in REFERRAL-SIZE-TOO-LARGE-1.

File: docs/public/specifications/test-zones/README.md [hunk 6]
```diff
@@ -126,6 +127,34 @@ owner names of such data must follow the reverse data standards. There is no
 requirements for creating separate zones for `in-addr.arpa` or `ip6.arpa` or
 below.
 
+### Undelegated data
+
+Some test scenarios require that the test zones is resolved as
+[undelegated test]. In that case, the scenario specification will contain a small
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/README.md:133]:
```suggestion
Some test scenarios require that an [undelegated test] be carried out on it.
In that case, the scenario specification will contain a small
```

File: docs/public/specifications/test-zones/Delegation-TP/delegation03.md (hunk 5 above)
Comment (matsduf) [reply docs/public/specifications/test-zones/Delegation-TP/delegation03.md:113]: Fixed.

**Comment by matsduf** [comment]
@marc-vanderwal, please re-review.

**Review by marc-vanderwal — APPROVED** [review]

**Comment by matsduf** [comment]
Resolved conflict that was created when #1294 was merged.

@tgreenx and @marc-vanderwal, please review/approve.

**Pull request merged**